### ISSUE TITLE
Defines: NO_SSL3_METHOD needs API_COMPAT?

### ISSUE THREAD

**Issue by traud** [openssl/openssl#6346]
I know, this is a mixture between support question and bug report. Even if turns out to be an usability bug, I guess, this gets a won’t-fix, because this issue exists since day one of OpenSSL 1.1.0. Anyway, because I am unsure, I am reporting this:

```Shell
$ ./config enable-ssl3-method no-deprecated
```

is a bit silly configuration. However currently, this is a fully valid configuration. OpenSSL users might have legacy code which still allows SSLv3-only (no discussion about this, please, it is not code under my control). For OpenSSL 1.1.x, those users have to adapt and change their source code:

```C
#if !defined(OPENSSL_NO_SSL2) && !(OPENSSL_VERSION_NUMBER >= 0x10100000L)
… = SSL_CTX_new(SSLv2_client_method());
#endif

#if !defined(OPENSSL_NO_SSL3_METHOD) && !(defined(OPENSSL_API_COMPAT) && (OPENSSL_API_COMPAT >= 0x10100000L))
… = SSL_CTX_new(SSLv3_client_method());
#endif
```

This is required to catch the above corner-case configuration. My questions:

1.  Should 'no-deprecated' not overrule 'enable-ssl3-method', which sets 'OPENSSL_NO_SSL3_METHOD' as it does with 'no-ssl3-method'?

2.  Should OpenSSL 1.1.x not define 'OPENSSL_NO_SSL2' as well?

3.  What was the intention to spilt OPENSSL_NO_SSLx and OPENSSL_NO_SSLx_METHOD exactly? Looking at the source code, none seems to overrule the other, they look independent. However, to configure with SSLv3, I have to enable both: `./config enable-ssl3 enable-ssl3-method`. Furthermore, as seen in the above code snippet, the Defines between SSLv2 and SSLv3 get asymmetric because SSlv2 has to be without `_METHOD`, SSlv3 has be with `_METHOD`. No idea how to repair this; just asking out of curiosity.

**Comment by richsalz** [comment]
This is reported against 1.1.0 and later, so I am removing the 1.1.1 milestone.

**Comment by t8m** [comment]
1. no-deprecated is not a production configuration (apart from very special use-cases) and yeah, it should not be used with enable-ssl3-method.
2. I do not think we should add OPENSSL_NO_SSL2 (or similar NO_xxxxxxx for deleted features). Using `OPENSSL_VERSION_NUMBER >=` for those is IMO appropriate.
3. the NO_SSLx_METHOD switches off the single version methods where the OPENSSL_NO_SSLx switches off the support in the wildcard method.

Closing. Please reopen or open new separate issues if there is still anything to resolve.

**Comment by traud** [comment]
Not sure if I am able/allowed to re-open issues. Anyway, let us try this way first:

> I do not think we should add OPENSSL_NO_SSL2 (or similar NO_xxxxxxx for deleted features). Using OPENSSL_VERSION_NUMBER >= for those is IMO appropriate.

Then, I do not understand the current OpenSSL approach. For example, `DH_free()` is deprecated since OpenSSL 3.0. I used that to offer Forward Secrecy based on Diffie–Hellman key exchange as TLS server (DHE). There is no replacement. The situation in future is like having OPENSSL_NO_DH. However, OPENSSL_NO_DH is not set in OpenSSL 3.0 when it was configured with no-deprecated. How do I #ifdef the import of DH parameters in an open-source project which has to compatible with a wide range of OpenSSL versions:

a) OPENSSL_VERSION_NUMBER; does not work for me.
b) OPENSSL_API_COMPAT or
c) OPENSSL_API_LEVEL; seems to be the favorite according to commit a6a4d0a.

This creates hard to read source code in contrast to a simply #ifndef NO_xxxxxxx (which is hard to read already because it is a double negative).

**Comment by t8m** [comment]
> For example, `DH_free()` is deprecated since OpenSSL 3.0. I used that to offer Forward Secrecy based on Diffie–Hellman key exchange as TLS server (DHE). There is no replacement.

Have you read the migration guide? There is of course non-deprecated replacement for DH_free() (all the DH_ functions in general). That is the EVP_PKEY API. This API is available in all supported OpenSSL releases (1.1.1 that is).
